lib/llvmopencl/UnreachablesToReturns.cc

```diff
@@ -138,106 +144,125 @@ static bool convertUnreachablesToReturns(Function &F) {
-// unreachable; if the predecessor has conditional branch, make it unconditional
-static void detachBBFromPredecessor(BasicBlock *BB,
-                                    SmallBBSet &UnreachableBBs2) {
-  if (BB->hasNPredecessors(0))
-    return;
+// Fix predecessor BBs of BB which has an unreachable terminator inst
+// to ignore the BB.
+//
+// If the predecessor has an unconditional branch, replaces the branch with
+// UnreachableInst and adds the BB to \p NewUnreachableBBs. If the predecessor
+// has a conditional branch, makes it unconditional. The function should
+// be called until NewUnreachableBBs is empty.
+static void detachBBFromPredecessor(BasicBlock &BB,
+                                    SmallBBSet &NewUnreachableBBs) {
+  if (BB.hasNPredecessors(0))
+    return; // Already handled.
+
+  LLVM_DEBUG(
+      dbgs() << "Detaching a BB that has an unreachable or leads to one:\n");
+  LLVM_DEBUG(BB.dump());
 
   // To avoid invalidating the predecessors iterator,
   // store replacement instructions and replace after the loop
   SmallMapVector<Instruction *, Instruction *, 8> Replacements;
 
-  for (BasicBlock *Pred : predecessors(BB)) {
-    assert(Pred);
+  // If the BB is in a switch case, this is the switch instruction to fix.
+  SwitchInst *PredSwitch = nullptr;
+
+  for (BasicBlock *Pred : predecessors(&BB)) {
+    assert(Pred != nullptr);
     Instruction *I = Pred->getTerminator();
-    assert(I);
+    assert(I != nullptr);
     if (BranchInst *BI = dyn_cast<BranchInst>(I)) {
       if (BI->isUnconditional()) {
-        detachBBFromPredecessor(Pred, UnreachableBBs2);
-        // predecessor is unconditional branch - remove this block too
-        UnreachableBBs2.insert(Pred);
-        // replace unconditional branch with unreachable BB
-        UnreachableInst *UI = new UnreachableInst(BB->getContext());
-        Replacements.insert(std::make_pair(BI, UI));
+        LLVM_DEBUG(
+            dbgs() << "The predecessor is unconditionally branching to it\n");
+        LLVM_DEBUG(Pred->dump());
+        // The predecessor has an unconditional branch to the unreachable BB,
+        // remove that in a next call.
+        NewUnreachableBBs.insert(Pred);
       } else {
-        // conditional branch. replace with unconditional branch to the other BB
+        LLVM_DEBUG(
+            dbgs() << "The predecessor is conditionally branching to it\n");
+        LLVM_DEBUG(Pred->dump());
+
         BasicBlock *Other = nullptr;
-        if (BI->getSuccessor(0) == BB)
+        if (BI->getSuccessor(0) == &BB)
           Other = BI->getSuccessor(1);
         else
           Other = BI->getSuccessor(0);
         BranchInst *NewBI = BranchInst::Create(Other);
         Replacements.insert(std::make_pair(BI, NewBI));
       }
+    } else if (PredSwitch = dyn_cast<SwitchInst>(I)) {
+      // The BB can be only in one switch...case, but "thanks" to gotos
+      // there could be also other predecessors.
```

Comment:
Not sure if this is true. The language reference talks about duplicate *entries* that may not appear in the table - does it means that duplicate *value and label pairs* are not allowed? 

Also, consider that the BB could be referenced through the default case for which the no-duplicate-entries rule probably doesn't apply.

Reply:
The comment was meant to refer to the whole switch construct itself from language perspective. But the handling of the SwitchInst was indeed insufficient, I expanded it below.

Reply:
> duplicate entries that may not appear in the table - does it means that duplicate value and label pairs are not allowed?

I didn't think this through. Of course it means that duplicate case values are not allowed.